src/cephadm/tests/test_agent.py

```diff
@@ -0,0 +1,544 @@
+from unittest import mock
+import copy, json, os, threading
+
+import pytest
+
+from tests.fixtures import with_cephadm_ctx, cephadm_fs, import_cephadm
+
+_cephadm = import_cephadm()
+
+
+FSID = "beefbeef-beef-beef-1234-beefbeefbeef"
+AGENT_ID = 'host1'
+AGENT_DIR = f'/var/lib/ceph/{FSID}/agent.{AGENT_ID}'
+
+
+def test_agent_validate():
+    required_files = _cephadm.CephadmAgent.required_files
+    with with_cephadm_ctx([]) as ctx:
+        agent = _cephadm.CephadmAgent(ctx, FSID, AGENT_ID)
+        for i in range(len(required_files)):
+            incomplete_files = {s: 'text' for s in [f for j, f in enumerate(required_files) if j != i]}
+            with pytest.raises(_cephadm.Error, match=f'required file missing from config: {required_files[i]}'):
+                agent.validate(incomplete_files)
+        all_files = {s: 'text' for s in required_files}
+        agent.validate(all_files)
+
+
+@mock.patch('cephadm.call_throws')
+def test_agent_deploy_daemon_unit(_call_throws, cephadm_fs):
+    _call_throws.return_value = ('', '', 0)
+    agent_id = AGENT_ID
+
+    def _check_file(path, content):
```

Comment:
This function doesn't look like a closure, so it might be better to just have it defined at the level of the file. Generally, when functions are defined inside other functions I expect them to be closures.